File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateOnce(ListNode* head)
    {
        ListNode* last = head->next;
        ListNode* prev = head;
        int len = 1;
        while(last->next)
        {
            last = last->next;
            prev = prev->next;
        }
        last->next = head;
        head = last;
        prev->next = NULL;
        return head;
    }
    ListNode* rotateRight(ListNode* head, int k) {
        if(!head)
        return NULL;
        if(head->next==NULL)
        {
            return head;
        }
        ListNode * curr = head;
        int len = 0;
        while(curr)
        {
            curr = curr->next;
            len++;
        }
        cout<<len;

        k = k%len;

        while(k>0)
        {
            head = rotateOnce(head);
            k--;
        }
        return head;
        
    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k ==0:
            return head
        n = 0 
        curr = head 
        while curr:
            n+=1
            curr = curr.next 
        if n<=1:
            return head
        k = k%n
        if k ==0:
            return head
        dummy = ListNode(0,head)
        slow = dummy 
        fast = head 
        for _ in range(k):
            fast = fast.next 
        while fast:
            fast = fast.next 
            slow = slow.next 
        end = slow 
        newhead = end.next 
        end.next = None 
        cur = newhead 
        while cur.next:
            cur = cur.next 
        cur.next = head 
        return newhead
        